### PR TITLE
feat(eslint): rule for @o3rWidget and @o3rWidgetParam tags

### PR DESCRIPTION
**Pull request by matthieu-crouzet**
## Proposed change
Rule for `@o3rWidget` and `@o3rWidgetParam` tags

### CODE REVIEW

File: packages/@o3r/eslint-plugin/src/rules/o3r-widget-tags/o3r-widget-tags.ts [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+        if (!comment || !comment.value.length) {
+          return;
+        }
+
+        const { loc, value: docText } = comment;
+        const widgetTypes = Array.from(docText.matchAll(/@o3rWidget (.*)/g));
+
+        if (widgetTypes.length > 1) {
+          return context.report({
+            messageId: 'duplicatedWidget',
```
Comment (cpaulve-1A) [inline packages/@o3r/eslint-plugin/src/rules/o3r-widget-tags/o3r-widget-tags.ts:73]: is that the widget is duplicated or that the feature does not support more than widget?

File: packages/@o3r/eslint-plugin/src/rules/o3r-widget-tags/o3r-widget-tags.ts [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+    },
+    schema: [
+      {
+        type: 'object',
+        additionalProperties: {
+          type: 'object',
+          additionalProperties: false,
+          patternProperties: {
+            [o3rWidgetParameterPattern]: {
+              type: 'string',
```
Comment (kpanot) [inline packages/@o3r/eslint-plugin/src/rules/o3r-widget-tags/o3r-widget-tags.ts:31]: Would be more interesting to be:

```json5
{
  "type": "object",
  "required": ["type"],
  "properties": {
    "required": { "type": "boolnea" },
    "type": {
      "type": "string",
      "enum": [/* ... */]
    }
  }
}
```

To indicate that the paremeter is mandatory (and then provided by the suggestion)

File: packages/@o3r/eslint-plugin/src/rules/o3r-widget-tags/o3r-widget-tags.ts [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+      duplicatedWidget: '@o3rWidget must be defined only once.'
+    }
+  },
+  defaultOptions: [],
+  create: (context) => {
+    const options: O3rWidgetTagsRuleOption = context.options.reduce((acc, option) => ({ ...acc, ...option }), {});
+    const supportedO3rWidgets = new Set(Object.keys(options));
+    return {
+      // eslint-disable-next-line @typescript-eslint/naming-convention
+      TSPropertySignature: (node) => {
```
Comment (kpanot) [inline packages/@o3r/eslint-plugin/src/rules/o3r-widget-tags/o3r-widget-tags.ts:60]: This should be restricted to properties in an interface extending `Configuration` (and maybe raise an error if it is not the case)

File: packages/@o3r/extractors/src/utils/config-doc.ts [hunk 4]
```diff
@@ -61,7 +61,7 @@ export function getWidgetInformationFromDocComment(docText: string): ConfigPrope
     return;
   }
 
-  const widgetParameters = Array.from(docText.matchAll(/@o3rWidgetParameter (.*)/g))
+  const widgetParameters = Array.from(docText.matchAll(/@o3rWidgetParam (.*)/g))
```
Comment (kpanot) [inline packages/@o3r/extractors/src/utils/config-doc.ts:64]: This why UTs can help sometime :D

File: packages/@o3r/eslint-plugin/src/rules/o3r-widget-tags/o3r-widget-tags.ts [hunk 5]
```diff
@@ -0,0 +1,168 @@
+import { createRule } from '../utils';
+
+const o3rWidgetParameterPattern = '^[a-zA-Z0-9-_:.]+$';
+
+export interface O3rWidgetTagsRuleOption {
+  [widgetName: string]: {
+    [paramName: string]: 'string' | 'number' | 'boolean' | 'string[]' | 'number[]' | 'boolean[]';
```
Comment (kpanot) [inline packages/@o3r/eslint-plugin/src/rules/o3r-widget-tags/o3r-widget-tags.ts:7]: Could be externalized and apply to the `enum` above to make sure they are aligned

File: packages/@o3r/eslint-plugin/src/rules/o3r-widget-tags/o3r-widget-tags.ts (hunk 2 above)
Comment (matthieu-crouzet) [reply packages/@o3r/eslint-plugin/src/rules/o3r-widget-tags/o3r-widget-tags.ts:31]: I had discussion yesterday with Guillaume(s) and any parameter from a widget will be required

File: packages/@o3r/eslint-plugin/src/rules/o3r-widget-tags/o3r-widget-tags.ts (hunk 1 above)
Comment (matthieu-crouzet) [reply packages/@o3r/eslint-plugin/src/rules/o3r-widget-tags/o3r-widget-tags.ts]: I change the `messageId` to `onlyOneWidgetAllowed`

File: packages/@o3r/eslint-plugin/src/rules/o3r-widget-tags/o3r-widget-tags.ts (hunk 2 above)
Comment (kpanot) [reply packages/@o3r/eslint-plugin/src/rules/o3r-widget-tags/o3r-widget-tags.ts:44]: It's more to configure rule to decide to make it required in the code than having required in the CMS

File: packages/@o3r/eslint-plugin/src/rules/o3r-widget-tags/o3r-widget-tags.ts [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+            }
+          }
+        }
+      }
+    ],
+    messages: {
+      notSupportedType: '{{ o3rWidgetType }} is not supported. Supported o3rWidget types: {{ supportedO3rWidgets }}.',
+      notSupportedParamForType: '{{ o3rWidgetParam }} is not supported for {{ o3rWidgetType }}. Supported o3rWidgetParam for {{ o3rWidgetType }}: {{ supportedO3rWidgetParam }}.',
+      invalidParamValueType: '{{ o3rWidgetParam }} supports only {{ o3rWidgetParamType }}.',
+      noParamWithoutWidget: '@o3rWidgetParam could not be used without @o3rWidget',
```
Comment (cpaulve-1A) [inline packages/@o3r/eslint-plugin/src/rules/o3r-widget-tags/o3r-widget-tags.ts:55]:
```suggestion
      noParamWithoutWidget: '@o3rWidgetParam can not be used without @o3rWidget',
```

File: packages/@o3r/eslint-plugin/src/rules/o3r-widget-tags/o3r-widget-tags.ts [hunk 7]
```diff
@@ -0,0 +1,303 @@
...
+          return;
+        }
+
+        const interfaceDeclNode = node.parent?.parent;
+        if (
+          interfaceDeclNode?.type !== TSESTree.AST_NODE_TYPES.TSInterfaceDeclaration
+          || !interfaceDeclNode.extends?.some((ext) =>
+            ext.type === TSESTree.AST_NODE_TYPES.TSInterfaceHeritage
+            && ext.expression.type === TSESTree.AST_NODE_TYPES.Identifier
+            && ['Configuration', 'NestedConfiguration', 'DynamicConfiguration'].includes(ext.expression.name)
```
Comment (kpanot) [inline packages/@o3r/eslint-plugin/src/rules/o3r-widget-tags/o3r-widget-tags.ts:151]: Not sure `DynamicConfiguration` exists.
Maybe `['Configuration', 'NestedConfiguration', 'DynamicConfiguration']` should be set as default value of a config to be overridable

File: packages/@o3r/eslint-plugin/src/rules/o3r-widget-tags/o3r-widget-tags.spec.ts [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+  /**
+   * @o3rWidget supportedType
+   * @o3rWidgetParam supportedParam "valid value"
+   */
+  prop: string;
+}
+`;
+
+const options: O3rWidgetTagsRuleOption[] = [{
+  widgets: {
```
Comment (cpaulve-1A) [inline packages/@o3r/eslint-plugin/src/rules/o3r-widget-tags/o3r-widget-tags.spec.ts:23]: You don't update the documentation? Isn't that the options for the rule defined in the rules.md?

**Review by cpaulve-1A — APPROVED** [review]

**Review by fpaul-1A — APPROVED** [review]

**Pull request merged**